x-pack/plugins/rule_registry/server/rule_data_plugin_service/rule_data_plugin_service.ts

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { ElasticsearchClient, Logger } from 'kibana/server';
+
+import { RuleDataClient } from '../rule_data_client';
+import { IndexNames } from './index_names';
+import { IndexOptions } from './index_options';
+import { ResourceNames } from './resource_names';
+import { ResourceInstaller, Resources } from './resource_installer';
+
+interface ConstructorOptions {
+  getClusterClient: () => Promise<ElasticsearchClient>;
+  logger: Logger;
+  isWriteEnabled: boolean;
+  index: string;
+}
+
+export class RuleDataPluginService {
+  private readonly resourceNames: ResourceNames;
+  private readonly resourceInstaller: ResourceInstaller;
+  private readonly installCommonResources: () => Promise<void>;
+
+  constructor(private readonly options: ConstructorOptions) {
+    this.resourceNames = new ResourceNames({ indexPrefixFromConfig: options.index });
+    this.resourceInstaller = new ResourceInstaller({
+      resourceNames: this.resourceNames,
+      getClusterClient: options.getClusterClient,
+      logger: options.logger,
+      isWriteEnabled: options.isWriteEnabled,
```

Comment:
Leaving this as a note for pass 2 or a follow up PR: if any of the resource installation routines fail, we should disable writes.

Reply:
Currently in `RuleDataWriter` it's

```ts
        if (!isWriteEnabled) {
          throw new RuleDataWriteDisabledError();
        }

        const { clusterClient } = await waitUntilIndexIsReady();
```

- clients using it have to check `RuleDataClient.isWriteEnabled()` before calling `IRuleDataWriter.bulk()`
- if any of the resource installation routines fail, it will throw on `await waitUntilIndexIsReady()`

Not sure about this DX, but it works. What was your idea regarding disabling writes?

Reply:
I like the idea of `waitUntilReadyForWriting` throwing if it fails to install the resources. Looks like it's been updated to return null, I think that will give us less useful errors - when `writer.bulk` fails we'll see something like `can't find property bulk of null` whereas if we re-throw in `installIndexResources` we can explain why the writer is refusing to write.

Reply:
> Looks like it's been updated to return null if it fails

@marshallmain sorry, I pushed unfinished code 🤦  I'm still trying to make the CI pass.
The issue with index bootstrapping promises was that in case of any exception during bootstrapping they were turning into unhandled promise rejections which was causing the shutdown of Kibana process.

I'm fixing that.

Reply:
> I like the idea of waitUntilReadyForWriting throwing if it fails to install the resources.

This should work now as expected.